File: problems/1137-n-th-tribonacci-number.ts
export {}

function tribonacciIterative(n: number): number {
  switch (n) {
    case 0:
      return 0
    case 1:
    case 2:
      return 1
    default:
      let a = 1
      let b = 1
      let c = 2
      for (let i = 3; i < n; i++) {
        let temp = a + b + c
        a = b
        b = c
        c = temp
      }
      return c
  }
}

// Recursive solution.
const memo = new Map<number, number>()
memo.set(0, 0)
memo.set(1, 1)
memo.set(2, 1)

function tribonacci(n: number): number {
  if (memo.has(n)) {
    return memo.get(n)!
  }
  switch (n) {
    case 0:
      return 0
    case 1:
    case 2:
      return 1
    default:
      const value = tribonacci(n - 1) + tribonacci(n - 2) + tribonacci(n - 3)
      memo.set(n, value)
      return value
  }
}

File: problems/1137_n_th_tribonacci_number.py
class Solution:
    memo: list[int | None]

    def __init__(self):
        self.memo = [0, 1, 1, 2, 4, 7, 13, 24] * 30

    def tribonacci(self, n: int) -> int:
        entry = self.memo[n]
        if entry is not None:
            return entry

        match n:
            case 0:
                ans = 0
            case 1:
                ans = 1
            case 2:
                ans = 1
            case _:
                ans = (
                    self.tribonacci(n - 1)
                    + self.tribonacci(n - 2)
                    + self.tribonacci(n - 3)
                )

        self.memo[n] = ans
        return ans
